# Logistic Regression - Breast Cancer Detection

In [2]:
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.model_selection import train_test_split
from sklearn.datasets import load_breast_cancer

In [3]:
# Load the dataset
data=load_breast_cancer(as_frame=True)

# convert it to a dataframe
df=data.frame
df

,mean radius,mean texture,mean perimeter,mean area,mean smoothness,mean compactness,mean concavity,mean concave points,mean symmetry,mean fractal dimension,...,worst texture,worst perimeter,worst area,worst smoothness,worst compactness,worst concavity,worst concave points,worst symmetry,worst fractal dimension,target
0,17.99,10.38,122.80,1001.0,0.11840,0.27760,0.30010,0.14710,0.2419,0.07871,...,17.33,184.60,2019.0,0.16220,0.66560,0.7119,0.2654,0.4601,0.11890,0
1,20.57,17.77,132.90,1326.0,0.08474,0.07864,0.08690,0.07017,0.1812,0.05667,...,23.41,158.80,1956.0,0.12380,0.18660,0.2416,0.1860,0.2750,0.08902,0
2,19.69,21.25,130.00,1203.0,0.10960,0.15990,0.19740,0.12790,0.2069,0.05999,...,25.53,152.50,1709.0,0.14440,0.42450,0.4504,0.2430,0.3613,0.08758,0
3,11.42,20.38,77.58,386.1,0.14250,0.28390,0.24140,0.10520,0.2597,0.09744,...,26.50,98.87,567.7,0.20980,0.86630,0.6869,0.2575,0.6638,0.17300,0
4,20.29,14.34,135.10,1297.0,0.10030,0.13280,0.19800,0.10430,0.1809,0.05883,...,16.67,152.20,1575.0,0.13740,0.20500,0.4000,0.1625,0.2364,0.07678,0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
564,21.56,22.39,142.00,1479.0,0.11100,0.11590,0.24390,0.13890,0.1726,0.05623,...,26.40,166.10,2027.0,0.14100,0.21130,0.4107,0.2216,0.2060,0.07115,0
565,20.13,28.25,131.20,1261.0,0.09780,0.10340,0.14400,0.09791,0.1752,0.05533,...,38.25,155.00,1731.0,0.11660,0.19220,0.3215,0.1628,0.2572,0.06637,0
566,16.60,28.08,108.30,858.1,0.08455,0.10230,0.09251,0.05302,0.1590,0.05648,...,34.12,126.70,1124.0,0.11390,0.30940,0.3403,0.1418,0.2218,0.07820,0
567,20.60,29.33,140.10,1265.0,0.11780,0.27700,0.35140,0.15200,0.2397,0.07016,...,39.42,184.60,1821.0,0.16500,0.86810,0.9387,0.2650,0.4087,0.12400,0


In [4]:
# Show the first 5 rows
df.head()

,mean radius,mean texture,mean perimeter,mean area,mean smoothness,mean compactness,mean concavity,mean concave points,mean symmetry,mean fractal dimension,...,worst texture,worst perimeter,worst area,worst smoothness,worst compactness,worst concavity,worst concave points,worst symmetry,worst fractal dimension,target
0,17.99,10.38,122.80,1001.0,0.11840,0.27760,0.3001,0.14710,0.2419,0.07871,...,17.33,184.60,2019.0,0.1622,0.6656,0.7119,0.2654,0.4601,0.11890,0
1,20.57,17.77,132.90,1326.0,0.08474,0.07864,0.0869,0.07017,0.1812,0.05667,...,23.41,158.80,1956.0,0.1238,0.1866,0.2416,0.1860,0.2750,0.08902,0
2,19.69,21.25,130.00,1203.0,0.10960,0.15990,0.1974,0.12790,0.2069,0.05999,...,25.53,152.50,1709.0,0.1444,0.4245,0.4504,0.2430,0.3613,0.08758,0
3,11.42,20.38,77.58,386.1,0.14250,0.28390,0.2414,0.10520,0.2597,0.09744,...,26.50,98.87,567.7,0.2098,0.8663,0.6869,0.2575,0.6638,0.17300,0
4,20.29,14.34,135.10,1297.0,0.10030,0.13280,0.1980,0.10430,0.1809,0.05883,...,16.67,152.20,1575.0,0.1374,0.2050,0.4000,0.1625,0.2364,0.07678,0


In [5]:
df.shape

(569, 31)

In [6]:
# Check for null values
df.isnull().sum()

mean radius                0
mean texture               0
mean perimeter             0
mean area                  0
mean smoothness            0
mean compactness           0
mean concavity             0
mean concave points        0
mean symmetry              0
mean fractal dimension     0
radius error               0
texture error              0
perimeter error            0
area error                 0
smoothness error           0
compactness error          0
concavity error            0
concave points error       0
symmetry error             0
fractal dimension error    0
worst radius               0
worst texture              0
worst perimeter            0
worst area                 0
worst smoothness           0
worst compactness          0
worst concavity            0
worst concave points       0
worst symmetry             0
worst fractal dimension    0
target                     0
dtype: int64

In [7]:
# Check for duplicates
duplicate_count=df.duplicated().sum()
print(f"Number of duplicate rows : {duplicate_count}")

Number of duplicate rows : 0


In [10]:
# Check outliers for 'mean radius' using IQR
Q1=df['mean radius'].quantile(0.25)
Q3=df['mean radius'].quantile(0.75)
IQR = Q3-Q1

#Outlier Range

lower=Q1-1.5*IQR
upper=Q3+1.5*IQR

outliers=df[(df['mean radius']<lower) | (df['mean radius']>upper)]
print(f"Number of outliers in mean radius : {outliers.shape[0]}")


Number of outliers in mean radius : 14


### These 14 data points have a 'mean radius' either too small or too large compared to most of the other values.

In a boxplot, they appear as dots outside the "whiskers" (the normal range).

## Visualize Outliers using BoxPlot

In [ ]:
sns.boxplot(df['mean radius'],color='skyblue')
plt.title("Outliers in mean radius")
plt.show()

In [ ]:
df.shape

In [11]:
# remove outliers in 'mean radius' only

clean_outliers=df[(df['mean radius']>=lower) & (df['mean radius']<=upper)]
print(f"New shape after removing outliers {clean_outliers.shape}")

New shape after removing outliers (555, 31)


## Split features & Target

In [ ]:
x=df.drop('target',axis=1)
y=df['target']

# 80% Train , 20% Test
x_train,x_test,y_train,y_test=train_test_split(x,y,test_size=0.2,random_state=42)

## Train Logistic Regression Model

In [ ]:
from sklearn.linear_model import LogisticRegression

#create model & fit

model=LogisticRegression(max_iter=10000)
model.fit(x_train,y_train)

In [ ]:
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report

# Predict on test data
y_pred=model.predict(x_test)

# Accuracy
print("Accuracy : ",accuracy_score(y_test,y_pred))
print("\n")

# Confusion matrix
print("Confusion Matrix :",confusion_matrix(y_test,y_pred))
print("\n")

# Classification Report
print("classification Report :",classification_report(y_test,y_pred))

Class 0 means the person is diagnosed with cancer

Class 1 means the person is not diagnosed with cancer

In [ ]:
# Generate Confusion Matrix
cm=confusion_matrix(y_test,y_pred)

#plot heatmap
sns.heatmap(cm,annot=True,fmt='d', cmap='Blues',xticklabels=['0', '1'], yticklabels=['0', '1'])
plt.xlabel("Predicted")
plt.ylabel("Actual")
plt.title("Confusion Metrics")
plt.show()

🔹 **Accuracy**: 0.9561 (≈ 96%)

This means the model correctly predicted whether the tumor was malignant or benign for 96% of the test cases.


🔹 **Confusion Matrix:**


Term	Meaning	Count


True Negative (TN)	Correctly predicted  	39  diagnosed with cancer

False Positive (FP)	Mistakenly predicted 4  diagnosed with cancer

False Negative (FN)	Mistakenly predicted  	1 not diagnosed with cancer

True Positive (TP)	Correctly predicted 70 not diagnosed with cancer


🔹 **Precision**:

Class 0 : 97%

Class 1 : 95%

Precision tells us when the model predicts a class, how often it's right.

🔹 **Recall**:

Class 0: 91%

Class 1: 99%

Recall tells us how well the model finds all positive cases.

🔹** F1-Score**:

Class 0: 0.94

Class 1: 0.97